SAMLFilterServletAdapterTest fails on Oracle

### Area

storage

### Describe the bug

For SAML, it runs queries on client attribute values. As those are CLOBs, a simple comparison on values doesn't work. 



### Version

20.0.1

### Expected behavior

Tests should be green

### Actual behavior

Test fail; here's just one: SAMLFilterServletAdapterTest.idpInitiatedLoginTest

```
11:05:09,917 INFO  [org.keycloak.testsuite.adapter.servlet.SAMLFilterServletAdapterTest] [SAMLFilterServletAdapterTest] salesPostEncSignedAssertionsAndDocumentTest() FINISHED


11:05:09,917 INFO  [org.keycloak.testsuite.adapter.servlet.SAMLFilterServletAdapterTest] [SAMLFilterServletAdapterTest] idpInitiatedLoginTest() STARTED
11:05:10,325 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
11:05:10,325 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  https://localhost:8543/auth/realms/demo/login-actions/authenticate?client_id=http%3A%2F%2Flocalhost%3A8280%2Fsales-post-enc%2F&tab_id=JOtmoYs8k9M
11:05:10,325 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to https://localhost:8543/auth/realms/demo/super-random-page
11:05:10,333 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  https://localhost:8543/auth/realms/demo/super-random-page
11:05:10,333 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
11:05:10,333 INFO  [org.keycloak.testsuite.adapter.servlet.SAMLFilterServletAdapterTest] deleting cookies in 'demo' realm
11:05:10,334 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
11:05:10,334 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  https://localhost:8543/auth/realms/demo/super-random-page
11:05:10,334 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to http://localhost:8280/employee2/
11:05:10,356 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  https://localhost:8543/auth/realms/demo/login-actions/authenticate?client_id=http%3A%2F%2Flocalhost%3A8280%2Femployee2%2F&tab_id=B-DsOElkXPc
11:05:10,356 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
In SendUsername Servlet doGet() check roles is true
In checkRoles() checking role manager for user bburke@redhat.com
Setting checkRoles to true
In SendUsername Servlet doPostElseWhere() - path: logout.jsp
In SendUsername Servlet doPost() check roles is true
11:05:10,422 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
11:05:10,422 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  http://localhost:8280/employee2/saml
11:05:10,423 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to http://localhost:8280/sales-post-sig-email/
11:05:10,451 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  https://localhost:8543/auth/realms/demo/login-actions/authenticate?client_id=http%3A%2F%2Flocalhost%3A8280%2Fsales-post-sig-email%2F&tab_id=qiWtOw9DDGQ
11:05:10,451 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
In SendUsername Servlet doGet() check roles is true
In checkRoles() checking role manager for user bburke@redhat.com
Setting checkRoles to true
In SendUsername Servlet doPostElseWhere() - path: logout.jsp
In SendUsername Servlet doPost() check roles is true
11:05:10,539 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
11:05:10,539 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  http://localhost:8280/sales-post-sig-email/saml
11:05:10,539 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to https://localhost:8543/auth/realms/demo/protocol/saml/clients/employee2
2022-11-22 11:05:10,546 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-0) SQL Error: 932, SQLState: 42000
2022-11-22 11:05:10,546 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-0) ORA-00932: inconsistent datatypes: expected - got NCLOB

2022-11-22 11:05:10,546 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-0) Uncaught server error: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:63)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:67)
	at org.hibernate.loader.Loader.getResultSet(Loader.java:2322)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2075)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037)
	at org.hibernate.loader.Loader.scroll(Loader.java:2945)
	at org.hibernate.loader.hql.QueryLoader.scroll(QueryLoader.java:610)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.scroll(QueryTranslatorImpl.java:452)
	at org.hibernate.engine.query.spi.HQLQueryPlan.performScroll(HQLQueryPlan.java:353)
	at org.hibernate.internal.SessionImpl.scroll(SessionImpl.java:1601)
	at org.hibernate.query.internal.AbstractProducedQuery.doScroll(AbstractProducedQuery.java:1588)
	at org.hibernate.query.internal.AbstractProducedQuery.scroll(AbstractProducedQuery.java:1574)
	at org.hibernate.query.internal.AbstractProducedQuery.stream(AbstractProducedQuery.java:1598)
	at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.stream(CriteriaQueryTypeQueryAdapter.java:91)
	at org.hibernate.query.Query.getResultStream(Query.java:1140)
	at org.keycloak.models.jpa.JpaRealmProvider.searchClientsByAttributes(JpaRealmProvider.java:804)
	at org.keycloak.storage.ClientStorageManager.lambda$searchClientsByAttributes$5(ClientStorageManager.java:167)
	at org.keycloak.storage.ClientStorageManager.query(ClientStorageManager.java:204)
	at org.keycloak.storage.ClientStorageManager.searchClientsByAttributes(ClientStorageManager.java:167)
	at org.keycloak.models.cache.infinispan.RealmCacheSession.searchClientsByAttributes(RealmCacheSession.java:1190)
	at org.keycloak.protocol.saml.SamlService.idpInitiatedSSO(SamlService.java:925)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:430)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:408)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:564)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.sql.SQLSyntaxErrorException: ORA-00932: inconsistent datatypes: expected - got NCLOB

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:630)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:564)
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1231)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:772)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:299)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:512)
	at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:163)
	at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1010)
	at oracle.jdbc.driver.OracleStatement.prepareDefineBufferAndExecute(OracleStatement.java:1271)
	at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:1149)
	at oracle.jdbc.driver.OracleStatement.executeSQLSelect(OracleStatement.java:1661)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1470)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3761)
	at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:3936)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1102)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeQuery(PreparedStatementWrapper.java:78)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:57)
	... 65 more
Caused by: Error : 932, Position : 206, Sql = select cliententi0_.ID as col_0_0_ from CLIENT cliententi0_ inner join CLIENT_ATTRIBUTES attributes1_ on cliententi0_.ID=attributes1_.CLIENT_ID where cliententi0_.REALM_ID=:1  and attributes1_.NAME=:2  and attributes1_.VALUE=:3  order by cliententi0_.CLIENT_ID asc fetch first :4  rows only, OriginalSql = select cliententi0_.ID as col_0_0_ from CLIENT cliententi0_ inner join CLIENT_ATTRIBUTES attributes1_ on cliententi0_.ID=attributes1_.CLIENT_ID where cliententi0_.REALM_ID=? and attributes1_.NAME=? and attributes1_.VALUE=? order by cliententi0_.CLIENT_ID asc fetch first ? rows only, Error Msg = ORA-00932: inconsistent datatypes: expected - got NCLOB

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:637)
	... 81 more

11:05:10,554 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  https://localhost:8543/auth/realms/demo/protocol/saml/clients/employee2
11:05:10,555 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
11:05:12,159 INFO  [org.keycloak.testsuite.adapter.servlet.SAMLFilterServletAdapterTest] calling all TestCleanup
```


### How to Reproduce?

Run the test on Oracle

### Anything else?

```
2022-11-22 17:24:52  [ERROR] org.keycloak.testsuite.admin.group.GroupTest.doNotAllowSameGroupNameAtTopLevelInDatabase  Time elapsed: 0.383 s  <<< FAILURE!
2022-11-22 17:24:52  java.lang.AssertionError: 
2022-11-22 17:24:52  
2022-11-22 17:24:52  Expected: (an instance of org.keycloak.testsuite.runonserver.RunOnServerException and exception with message a string containing "org.keycloak.models.ModelDuplicateException")
2022-11-22 17:24:52       but: an instance of org.keycloak.testsuite.runonserver.RunOnServerException <java.lang.RuntimeException: java.lang.ClassNotFoundException: oracle.jdbc.OracleDatabaseException> is a java.lang.RuntimeException
2022-11-22 17:24:52  Stacktrace was: java.lang.RuntimeException: java.lang.ClassNotFoundException: oracle.jdbc.OracleDatabaseException
2022-11-22 17:24:52  	at org.keycloak.testsuite.runonserver.SerializationUtil.decodeException(SerializationUtil.java:67)
2022-11-22 17:24:52  	at org.keycloak.testsuite.client.KeycloakTestingClient$Server.run(KeycloakTestingClient.java:156)
2022-11-22 17:24:52  	at org.keycloak.testsuite.admin.group.GroupTest.doNotAllowSameGroupNameAtTopLevelInDatabase(GroupTest.java:287)
2022-11-22 17:24:52  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-11-22 17:24:52  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-11-22 17:24:52  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-11-22 17:24:52  	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2022-11-22 17:24:52  	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2022-11-22 17:24:52  	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
```

--- a/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
+++ b/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java
@@ -40,7 +40,7 @@ import javax.persistence.criteria.Join;
 import javax.persistence.criteria.Predicate;
 import javax.persistence.criteria.Root;
 
-import org.apache.commons.lang.BooleanUtils;
+import org.hibernate.Session;
 import org.jboss.logging.Logger;
 import org.keycloak.common.util.Time;
 import org.keycloak.connections.jpa.util.JpaUtils;
@@ -788,6 +788,9 @@ public class JpaRealmProvider implements RealmProvider, ClientProvider, ClientSc
 
         predicates.add(builder.equal(root.get("realmId"), realm.getId()));
 
+        //noinspection resource
+        String dbProductName = em.unwrap(Session.class).doReturningWork(connection -> connection.getMetaData().getDatabaseProductName());
+
         for (Map.Entry<String, String> entry : filteredAttributes.entrySet()) {
             String key = entry.getKey();
             String value = entry.getValue();
@@ -795,7 +798,19 @@ public class JpaRealmProvider implements RealmProvider, ClientProvider, ClientSc
             Join<ClientEntity, ClientAttributeEntity> attributeJoin = root.join("attributes");
 
             Predicate attrNamePredicate = builder.equal(attributeJoin.get("name"), key);
-            Predicate attrValuePredicate = builder.equal(attributeJoin.get("value"), value);
+
+            Predicate attrValuePredicate;
+            if (dbProductName.equals("Oracle")) {
+                // SELECT * FROM client_attributes WHERE ... DBMS_LOB.COMPARE(value, '0') = 0 ...;
+                // Oracle is not able to compare a CLOB with a VARCHAR unless it being converted with TO_CHAR
+                // But for this all values in the table need to be smaller than 4K, otherwise the cast will fail with
+                // "ORA-22835: Buffer too small for CLOB to CHAR" (even if it is in another row).
+                // This leaves DBMS_LOB.COMPARE as the option to compare the CLOB with the value.
+                attrValuePredicate = builder.equal(builder.function("DBMS_LOB.COMPARE", Integer.class, attributeJoin.get("value"), builder.literal(value)), 0);
+            } else {
+                attrValuePredicate = builder.equal(attributeJoin.get("value"), value);
+            }
+
             predicates.add(builder.and(attrNamePredicate, attrValuePredicate));
         }
 

